File: llama_mobile-Android/src/main/java/com/llamamobile/LlamaMobile.kt
package com.llamamobile

import android.content.Context
import android.content.res.AssetManager
import java.io.IOException
import java.io.InputStream

/**
 * LlamaMobile Android Library
 * 
 * This class provides a Kotlin wrapper around the llama_mobile C library, 
 * allowing Android applications to interact with llama models.
 */
object LlamaMobile {
    
    /**
     * Available built-in grammar file names
     */
    enum class GrammarName {
        arithmetic,
        c,
        chess,
        english,
        japanese,
        json,
        json_arr,
        list
    }
    
    /**
     * Cache type enum
     */
    enum class CacheType {
        NONE,
        MEMORY
    }
    
    /**
     * Initialization parameters for creating a llama context
     * 
     * @property modelPath Path to the llama model file
     * @property chatTemplate Chat template to use (optional)
     * @property systemPrompt System prompt to guide the model's behavior (optional)
     * @property nCtx Size of the context window (default: 2048)
     * @property nBatch Batch size for processing input tokens (default: 512)
     * @property nUbatch Micro-batch size for processing input tokens (default: 512)
     * @property nGpuLayers Number of layers to offload to GPU (0 = no GPU acceleration)
     * @property nThreads Number of CPU threads to use for inference (default: 4)
     * @property useMmap Whether to use memory-mapped files for model loading (default: true)
     * @property useMlock Whether to lock model memory in RAM (prevents swapping, default: false)
     * @property embedding Whether to enable embedding generation (default: false)
     * @property poolingType Pooling type for embeddings (0 = no pooling, 1 = mean pooling, 2 = max pooling)
     * @property embdNormalize Whether to normalize embeddings (default: 0)
     * @property flashAttn Whether to enable flash attention optimization (default: false)
     * @property cacheTypeK Cache type for key tensors (e.g., "f16", "q4_0", optional)
     * @property cacheTypeV Cache type for value tensors (e.g., "f16", "q4_0", optional)
     */
    data class InitParams(
        val modelPath: String,
        val chatTemplate: String? = null,
        val systemPrompt: String? = null,
        val nCtx: Int = 2048,
        val nBatch: Int = 512,
        val nUbatch: Int = 512,
        val nGpuLayers: Int = 0,
        val nThreads: Int = 4,
        val useMmap: Boolean = true,
        val useMlock: Boolean = false,
        val embedding: Boolean = false,
        val poolingType: Int = 0,
        val embdNormalize: Int = 0,
        val flashAttn: Boolean = false,
        val cacheTypeK: String? = null,
        val cacheTypeV: String? = null
    )
    
    /**
     * Callback type for token streaming during completion generation
     */
    fun interface TokenCallback {
        /**
         * Called when a new token is generated
         * 
         * @param token The generated token string, or null to indicate end of generation
         * @return true to continue generation, false to stop
         */
        fun onToken(token: String?): Boolean
    }
    
    /**
     * Result structure for text completion generation
     * 
     * @property text Generated text completion
     * @property tokensPredicted Number of tokens predicted in the completion
     * @property tokensEvaluated Number of tokens evaluated from the prompt
     * @property truncated Whether the output was truncated due to context limits
     * @property stoppedEos Whether generation stopped at the end-of-sequence token
     * @property stoppedWord Whether generation stopped at a stop sequence
     * @property stoppedLimit Whether generation stopped due to reaching maximum tokens limit
     * @property stoppingWord The stop sequence that caused generation to stop (if any)
     */
    data class CompletionResult(
        val text: String,
        val tokensPredicted: Int,
        val tokensEvaluated: Int,
        val truncated: Boolean,
        val stoppedEos: Boolean,
        val stoppedWord: Boolean,
        val stoppedLimit: Boolean,
        val stoppingWord: String?
    )

    /**
     * Completion parameters for generating text
     * 
     * @property prompt Input prompt for text generation
     * @property nPredict Maximum number of tokens to generate (default: 128)
     * @property nThreads Number of CPU threads to use for generation (default: 4)
     * @property seed Random seed for generation (use -1 for random seed, default: -1)
     * @property temperature Temperature for sampling (higher values = more random output, default: 0.8)
     * @property topK Top-K sampling parameter (0 = disable, default: 40)
     * @property topP Top-P sampling parameter (nucleus sampling, default: 0.9)
     * @property minP Minimum probability for sampling (default: 0.05)
     * @property typicalP Typical-P sampling parameter (default: 1.0)
     * @property penaltyLastN Number of tokens to consider for repetition penalty (default: 64)
     * @property penaltyRepeat Penalty for repeated tokens (default: 1.1)
     * @property penaltyFreq Frequency penalty for tokens (default: 0.0)
     * @property penaltyPresent Present penalty for tokens (default: 0.0)
     * @property mirostat Mirostat sampling mode (0 = disable, 1 = Mirostat, 2 = Mirostat 2.0, default: 0)
     * @property mirostatTau Mirostat target entropy (default: 5.0)
     * @property mirostatEta Mirostat learning rate (default: 0.1)
     * @property ignoreEos Whether to ignore the end-of-sequence token (default: false)
     * @property nProbs Number of top probabilities to return per token (0 = disable, default: 0)
     * @property stopSequences Sequences that will stop generation when encountered (default: empty)
     * @property grammar Grammar for constrained generation (using GBNF format, optional)
     * @property tokenCallback Callback for streaming token generation (optional)
     */
    data class CompletionParams(
        val prompt: String,
        val nPredict: Int = 128,
        val nThreads: Int = 4,
        val seed: Int = -1,
        val temperature: Double = 0.8,
        val topK: Int = 40,
        val topP: Double = 0.9,
        val minP: Double = 0.05,
        val typicalP: Double = 1.0,
        val penaltyLastN: Int = 64,
        val penaltyRepeat: Double = 1.1,
        val penaltyFreq: Double = 0.0,
        val penaltyPresent: Double = 0.0,
        val mirostat: Int = 0,
        val mirostatTau: Double = 5.0,
        val mirostatEta: Double = 0.1,
        val ignoreEos: Boolean = false,
        val nProbs: Int = 0,
        val stopSequences: List<String> = emptyList(),
        val grammar: String? = null,
        val tokenCallback: TokenCallback? = null
    )
    
    /**
     * Returns the content of a built-in grammar file
     * 
     * @param context Application context to access assets
     * @param name The name of the grammar file without extension
     * @return The content of the grammar file, or null if not found or an error occurred
     */
    fun grammarContent(context: Context, name: GrammarName): String? {
        val assetManager: AssetManager = context.assets
        val fileName = "grammars/${name.name}.gbnf"
        
        return try {
            val inputStream: InputStream = assetManager.open(fileName)
            inputStream.bufferedReader().use { it.readText() }
        } catch (e: IOException) {
            e.printStackTrace()
            null
        }
    }
    
    /**
     * Loads the native libraries
     */
    init {
        System.loadLibrary("llama_mobile")
        System.loadLibrary("llama_mobile_jni")
    }
    
    /**
     * Initializes a new llama context
     * 
     * @param params Initialization parameters
     * @return Context handle, or 0 if initialization failed
     */
    external fun initContext(params: InitParams): Long
    
    /**
     * Generates text completion
     * 
     * @param contextHandle Context handle obtained from initContext
     * @param params Completion parameters
     * @return Generated text and metadata, or null if generation failed
     */
    external fun generateCompletion(contextHandle: Long, params: CompletionParams): CompletionResult?
    
    /**
     * Releases a llama context
     * 
     * @param contextHandle Context handle obtained from initContext
     */
    external fun releaseContext(contextHandle: Long)

    /**
     * Initializes a vocoder model for TTS
     * 
     * @param contextHandle Context handle obtained from initContext
     * @param vocoderModelPath Path to the vocoder model file
     * @return 0 if successful, non-zero otherwise
     */
    external fun initVocoder(contextHandle: Long, vocoderModelPath: String): Int

    /**
     * Checks if the vocoder is enabled
     * 
     * @param contextHandle Context handle obtained from initContext
     * @return true if vocoder is enabled, false otherwise
     */
    external fun isVocoderEnabled(contextHandle: Long): Boolean

    /**
     * Gets the TTS type
     * 
     * @param contextHandle Context handle obtained from initContext
     * @return TTS type identifier
     */
    external fun getTtsType(contextHandle: Long): Int

    /**
     * Gets formatted audio completion for TTS
     * 
     * @param contextHandle Context handle obtained from initContext
     * @param speakerJsonStr JSON string with speaker configuration
     * @param textToSpeak Text to convert to speech
     * @return Formatted audio completion string, or null if failed
     */
    external fun getFormattedAudioCompletion(contextHandle: Long, speakerJsonStr: String, textToSpeak: String): String?

    /**
     * Gets audio guide tokens for TTS
     * 
     * @param contextHandle Context handle obtained from initContext
     * @param textToSpeak Text to convert to speech
     * @return Array of audio guide tokens
     */
    external fun getAudioGuideTokens(contextHandle: Long, textToSpeak: String): IntArray?

    /**
     * Decodes audio tokens to float audio data
     * 
     * @param contextHandle Context handle obtained from initContext
     * @param tokens Audio tokens to decode
     * @return Decoded float audio data, or null if failed
     */
    external fun decodeAudioTokens(contextHandle: Long, tokens: IntArray): FloatArray?

    /**
     * Releases the vocoder resources
     * 
     * @param contextHandle Context handle obtained from initContext
     */
    external fun releaseVocoder(contextHandle: Long)
}

File: llama_mobile-capacitor-plugin/ios/Sources/LlamaMobilePlugin/LlamaMobile.swift
import Foundation
import llama_mobile

@objc public class LlamaMobile: NSObject {
    private var context: llama_mobile_context_handle_t?
    private var grammarsPath: String?
    private var isMultimodalInitialized: Bool = false
    
    @objc override public init() {
        super.init()
        // Set up grammars path if needed
        if let bundle = Bundle(for: type(of: self)) {
            grammarsPath = bundle.path(forResource: "grammars", ofType: nil)
        }
    }
    
    @objc public func initialize(_ params: [String: Any]) -> Bool {
        guard let modelPath = params["modelPath"] as? String else {
            return false
        }
        
        // Set up initialization parameters
        var initParams = llama_mobile_init_params_c_t()
        initParams.model_path = strdup(modelPath)
        
        // Set default values
        initParams.n_ctx = params["nCtx"] as? Int ?? 2048
        initParams.n_gpu_layers = params["nGpuLayers"] as? Int ?? 0
        initParams.n_threads = params["nThreads"] as? Int ?? 4
        initParams.n_batch = params["nBatch"] as? Int ?? 512
        initParams.n_ubatch = params["nUbatch"] as? Int ?? 512
        initParams.use_mmap = params["useMmap"] as? Bool ?? true
        initParams.use_mlock = params["useMlock"] as? Bool ?? false
        initParams.embedding = params["embedding"] as? Bool ?? false
        initParams.pooling_type = params["poolingType"] as? Int ?? 0
        initParams.embd_normalize = params["embdNormalize"] as? Bool ?? false
        initParams.flash_attn = params["flashAttn"] as? Bool ?? false
        
        // Optional parameters
        if let chatTemplate = params["chatTemplate"] as? String {
            initParams.chat_template = strdup(chatTemplate)
        }
        if let systemPrompt = params["systemPrompt"] as? String {
            initParams.system_prompt = strdup(systemPrompt)
        }
        if let cacheTypeK = params["cacheTypeK"] as? String {
            initParams.cache_type_k = strdup(cacheTypeK)
        }
        if let cacheTypeV = params["cacheTypeV"] as? String {
            initParams.cache_type_v = strdup(cacheTypeV)
        }
        
        // Initialize the context
        context = llama_mobile_init_context_c(&initParams)
        
        // Free allocated strings
        free(UnsafeMutableRawPointer(mutating: initParams.model_path))
        if initParams.chat_template != nil {
            free(UnsafeMutableRawPointer(mutating: initParams.chat_template))
        }
        if initParams.system_prompt != nil {
            free(UnsafeMutableRawPointer(mutating: initParams.system_prompt))
        }
        if initParams.cache_type_k != nil {
            free(UnsafeMutableRawPointer(mutating: initParams.cache_type_k))
        }
        if initParams.cache_type_v != nil {
            free(UnsafeMutableRawPointer(mutating: initParams.cache_type_v))
        }
        
        return context != nil
    }
    
    @objc public func generate(_ params: [String: Any]) -> [String: Any]? {
        guard let context = context, let prompt = params["prompt"] as? String else {
            return nil
        }
        
        // Set up completion parameters
        var completionParams = llama_mobile_completion_params_c_t()
        completionParams.prompt = strdup(prompt)
        
        // Set default values
        completionParams.max_tokens = params["maxTokens"] as? Int ?? 100
        completionParams.temperature = params["temperature"] as? Double ?? 0.8
        completionParams.top_k = params["topK"] as? Int ?? 40
        completionParams.top_p = params["topP"] as? Double ?? 0.95
        completionParams.min_p = params["minP"] as? Double ?? 0.05
        completionParams.typical_p = params["typicalP"] as? Double ?? 1.0
        completionParams.seed = params["seed"] as? Int ?? -1
        completionParams.n_threads = params["nThreads"] as? Int ?? 4
        completionParams.penalty_last_n = params["penaltyLastN"] as? Int ?? 64
        completionParams.penalty_repeat = params["penaltyRepeat"] as? Double ?? 1.1
        completionParams.penalty_freq = params["penaltyFreq"] as? Double ?? 0.0
        completionParams.penalty_present = params["penaltyPresent"] as? Double ?? 0.0
        completionParams.mirostat = params["mirostat"] as? Int ?? 0
        completionParams.mirostat_tau = params["mirostatTau"] as? Double ?? 5.0
        completionParams.mirostat_eta = params["mirostatEta"] as? Double ?? 0.1
        completionParams.ignore_eos = params["ignoreEos"] as? Bool ?? false
        
        // Optional grammar
        if let grammar = params["grammar"] as? String {
            completionParams.grammar = strdup(grammar)
        }
        
        // Optional stop sequences
        if let stopSequences = params["stopSequences"] as? [String], !stopSequences.isEmpty {
            let stopSequencesC = stopSequences.map { strdup($0) }
            completionParams.stop_sequences = UnsafeMutablePointer(mutating: stopSequencesC)
            completionParams.n_stop_sequences = stopSequencesC.count
        }
        
        // Generate completion
        let result = llama_mobile_completion_c(context, &completionParams)
        
        // Free allocated strings
        free(UnsafeMutableRawPointer(mutating: completionParams.prompt))
        if completionParams.grammar != nil {
            free(UnsafeMutableRawPointer(mutating: completionParams.grammar))
        }
        if completionParams.stop_sequences != nil {
            for i in 0..<completionParams.n_stop_sequences {
                free(UnsafeMutableRawPointer(mutating: completionParams.stop_sequences[i]))
            }
        }
        
        guard let resultStr = result else {
            return nil
        }
        
        let output = String(cString: resultStr)
        free(UnsafeMutableRawPointer(mutating: resultStr))
        
        // Create completion result
        return [
            "output": output,
            "tokensGenerated": 0, // Will be updated with actual value when available
            "tokensEvaluated": 0, // Will be updated with actual value when available
            "truncated": false,
            "stoppedEos": false,
            "stoppedWord": false,
            "stoppedLimit": false
        ]
    }
    
    @objc public func multimodalCompletion(_ params: [String: Any], mediaPaths: [String]) -> [String: Any]? {
        guard let context = context, let prompt = params["prompt"] as? String else {
            return nil
        }
        
        // Set up completion parameters
        var completionParams = llama_mobile_completion_params_c_t()
        completionParams.prompt = strdup(prompt)
        
        // Set default values
        completionParams.max_tokens = params["maxTokens"] as? Int ?? 100
        completionParams.temperature = params["temperature"] as? Double ?? 0.8
        completionParams.top_k = params["topK"] as? Int ?? 40
        completionParams.top_p = params["topP"] as? Double ?? 0.95
        completionParams.min_p = params["minP"] as? Double ?? 0.05
        completionParams.typical_p = params["typicalP"] as? Double ?? 1.0
        completionParams.seed = params["seed"] as? Int ?? -1
        completionParams.n_threads = params["nThreads"] as? Int ?? 4
        completionParams.penalty_last_n = params["penaltyLastN"] as? Int ?? 64
        completionParams.penalty_repeat = params["penaltyRepeat"] as? Double ?? 1.1
        completionParams.penalty_freq = params["penaltyFreq"] as? Double ?? 0.0
        completionParams.penalty_present = params["penaltyPresent"] as? Double ?? 0.0
        completionParams.mirostat = params["mirostat"] as? Int ?? 0
        completionParams.mirostat_tau = params["mirostatTau"] as? Double ?? 5.0
        completionParams.mirostat_eta = params["mirostatEta"] as? Double ?? 0.1
        completionParams.ignore_eos = params["ignoreEos"] as? Bool ?? false
        
        // Optional grammar
        if let grammar = params["grammar"] as? String {
            completionParams.grammar = strdup(grammar)
        }
        
        // Convert media paths to C array
        let mediaPathsC = mediaPaths.map { strdup($0) }
        
        // Generate multimodal completion
        let result = llama_mobile_multimodal_completion_c(context, &completionParams, mediaPathsC, mediaPathsC.count)
        
        // Free allocated strings
        free(UnsafeMutableRawPointer(mutating: completionParams.prompt))
        if completionParams.grammar != nil {
            free(UnsafeMutableRawPointer(mutating: completionParams.grammar))
        }
        for path in mediaPathsC {
            free(UnsafeMutableRawPointer(mutating: path))
        }
        
        guard let resultStr = result else {
            return nil
        }
        
        let output = String(cString: resultStr)
        free(UnsafeMutableRawPointer(mutating: resultStr))
        
        // Create completion result
        return [
            "output": output,
            "tokensGenerated": 0,
            "tokensEvaluated": 0,
            "truncated": false,
            "stoppedEos": false,
            "stoppedWord": false,
            "stoppedLimit": false
        ]
    }
    
    @objc public func stopCompletion() {
        guard let context = context else {
            return
        }
        llama_mobile_stop_completion_c(context)
    }
    
    @objc public func tokenize(_ text: String) -> [Int]? {
        guard let context = context else {
            return nil
        }
        
        let textC = strdup(text)
        defer { free(UnsafeMutableRawPointer(mutating: textC)) }
        
        var tokens: UnsafeMutablePointer<Int32>?
        var n_tokens: Int32 = 0
        
        if llama_mobile_tokenize_c(context, textC, &tokens, &n_tokens) == 0 {
            guard tokens != nil, n_tokens > 0 else {
                return nil
            }
            
            var result: [Int] = []
            for i in 0..<n_tokens {
                result.append(Int(tokens![i]))
            }
            
            free(UnsafeMutableRawPointer(mutating: tokens))
            return result
        }
        
        return nil
    }
    
    @objc public func detokenize(_ tokens: [Int]) -> String? {
        guard let context = context, !tokens.isEmpty else {
            return nil
        }
        
        let tokensInt32 = tokens.map { Int32($0) }
        let tokensC = UnsafeMutablePointer(mutating: tokensInt32)
        
        let result = llama_mobile_detokenize_c(context, tokensC, tokens.count)
        
        guard let resultStr = result else {
            return nil
        }
        
        let text = String(cString: resultStr)
        free(UnsafeMutableRawPointer(mutating: resultStr))
        
        return text
    }
    
    @objc public func generateEmbeddings(_ text: String) -> [Double]? {
        guard let context = context else {
            return nil
        }
        
        let textC = strdup(text)
        defer { free(UnsafeMutableRawPointer(mutating: textC)) }
        
        var embeddings: UnsafeMutablePointer<Float>?
        var n_embeddings: Int32 = 0
        
        if llama_mobile_generate_embeddings_c(context, textC, &embeddings, &n_embeddings) == 0 {
            guard embeddings != nil, n_embeddings > 0 else {
                return nil
            }
            
            var result: [Double] = []
            for i in 0..<n_embeddings {
                result.append(Double(embeddings![i]))
            }
            
            free(UnsafeMutableRawPointer(mutating: embeddings))
            return result
        }
        
        return nil
    }
    
    @objc public func applyLoraAdapters(_ adapters: [[String: Any]]) -> Bool {
        guard let context = context, !adapters.isEmpty else {
            return false
        }
        
        var loraParams: [llama_mobile_lora_adapter_c_t] = []
        
        for adapter in adapters {
            guard let path = adapter["path"] as? String, let scale = adapter["scale"] as? Double else {
                continue
            }
            
            var param = llama_mobile_lora_adapter_c_t()
            param.path = strdup(path)
            param.scale = Float(scale)
            loraParams.append(param)
        }
        
        let result = llama_mobile_apply_lora_adapters_c(context, loraParams, loraParams.count)
        
        // Free allocated paths
        for param in loraParams {
            free(UnsafeMutableRawPointer(mutating: param.path))
        }
        
        return result == 0
    }
    
    @objc public func removeLoraAdapters() {
        guard let context = context else {
            return
        }
        llama_mobile_remove_lora_adapters_c(context)
    }
    
    @objc public func initMultimodal(_ mmprojPath: String, useGpu: Bool) -> Bool {
        guard let context = context else {
            return false
        }
        
        let mmprojPathC = strdup(mmprojPath)
        defer { free(UnsafeMutableRawPointer(mutating: mmprojPathC)) }
        
        let result = llama_mobile_init_multimodal_c(context, mmprojPathC, useGpu)
        isMultimodalInitialized = result == 0
        return isMultimodalInitialized
    }
    
    @objc public func isMultimodalEnabled() -> Bool {
        return isMultimodalInitialized
    }
    
    @objc public func releaseMultimodal() {
        guard let context = context else {
            return
        }
        llama_mobile_release_multimodal_c(context)
        isMultimodalInitialized = false
    }
    
    @objc public func generateResponse(_ userMessage: String, maxTokens: Int) -> [String: Any]? {
        guard let context = context else {
            return nil
        }
        
        let userMessageC = strdup(userMessage)
        defer { free(UnsafeMutableRawPointer(mutating: userMessageC)) }
        
        let result = llama_mobile_generate_response_c(context, userMessageC, maxTokens)
        
        guard let resultStr = result else {
            return nil
        }
        
        let text = String(cString: resultStr)
        free(UnsafeMutableRawPointer(mutating: resultStr))
        
        // Create conversation result
        return [
            "text": text,
            "timeToFirstToken": 0,
            "totalTime": 0,
            "tokensGenerated": 0
        ]
    }
    
    @objc public func clearConversation() {
        guard let context = context else {
            return
        }
        llama_mobile_clear_conversation_c(context)
    }
    
    @objc public func getGrammarContent(_ grammarName: String) -> String? {
        guard let grammarsPath = grammarsPath else {
            return nil
        }
        
        let grammarFileName: String
        switch grammarName {
        case "arithmetic":
            grammarFileName = "arithmetic.gbnf"
        case "c":
            grammarFileName = "c.gbnf"
        case "chess":
            grammarFileName = "chess.gbnf"
        case "english":
            grammarFileName = "english.gbnf"
        case "japanese":
            grammarFileName = "japanese.gbnf"
        case "json":
            grammarFileName = "json.gbnf"
        case "jsonArr":
            grammarFileName = "json_arr.gbnf"
        case "list":
            grammarFileName = "list.gbnf"
        default:
            return nil
        }
        
        let grammarPath = "\(grammarsPath)/\(grammarFileName)"
        do {
            return try String(contentsOfFile: grammarPath, encoding: .utf8)
        } catch {
            return nil
        }
    }
    
    @objc public func release() {
        if let context = context {
            llama_mobile_release_context_c(context)
            self.context = nil
        }
        isMultimodalInitialized = false
    }
}
